cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/LocalizationUtils.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime;
+
+import co.cask.cdap.internal.app.runtime.distributed.LocalizeResource;
+import com.google.common.io.Files;
+import com.google.common.io.InputSupplier;
+import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
+import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Enumeration;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Utilities for file localization.
+ */
+public final class LocalizationUtils {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalizationUtils.class);
+
+  /**
+   * Localizes the specified {@link LocalizeResource} in the specified {@link File targetDir} with the specified
+   * file name and returns the {@link File} pointing to the localized file.
+   *
+   * @param fileName the name to localize the file with
+   * @param resource the {@link LocalizeResource} to localize
+   * @param targetDir the directory to localize the resource in
+   * @return the {@link File} pointing to the localized file.
+   */
+  public static File localizeResource(String fileName, LocalizeResource resource, File targetDir) throws IOException {
+    File tempFile = new File(targetDir, fileName);
+    File fileToLocalize = new File(resource.getURI().getPath());
+    if (resource.isArchive()) {
+      LOG.debug("Uncompress file {} to {}", fileToLocalize, tempFile);
+      uncompress(fileToLocalize, tempFile);
+    } else {
+      LOG.debug("Copy file from {} to {}", fileToLocalize, tempFile);
+      Files.copy(fileToLocalize, tempFile);
```

Comment:
Instead of copying, better try to do a hard link first. Only copy if fails.

http://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createLink(java.nio.file.Path,%20java.nio.file.Path)